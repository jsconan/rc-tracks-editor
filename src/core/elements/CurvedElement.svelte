<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { curvedElementPath } from '../fragments';
    import Shape from './Shape.svelte';

    export let cx = 0;
    export let cy = 0;
    export let width = 1;
    export let radius = 1;
    export let angle = 1;
    export let start = 0;
    export let d = 0;
    export let fill = void 0;
    export let stroke = void 0;
    export let transform = void 0;
</script>

<Shape path={curvedElementPath(cx, cy, width, radius, angle, start, d)} {fill} {stroke} {transform} />
