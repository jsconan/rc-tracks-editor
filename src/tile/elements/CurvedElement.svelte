<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { Vector2D } from '../../core/models';
    import { arcTo, attributeList, enlargeArc, lineTo, moveTo, RIGHT_ANGLE } from '../../core/helpers';

    export let cx = 0;
    export let cy = 0;
    export let width = 1;
    export let radius = 1;
    export let angle = 1;
    export let start = 0;
    export let d = 0;
    export let fill = void 0;
    export let stroke = void 0;
    export let transform = void 0;

    /**
     * Builds the SVG path for rendering a curve.
     * @param {number} curveRadius - The radius of the inner curve.
     * @param {number} curveWidth - The width between the inner and the outer curve.
     * @param {number} curveAngle - Tha angle of the curve.
     * @param {number} startAngle - The start angle of the curve.
     * @param {number} centerX - The X-coordinate of the center of the curve.
     * @param {number} centerY - The Y-coordinate of the center of the curve.
     * @param {number} addition - An additional distance added to the outline.
     * @private
     */
    function curvedElementPath(curveRadius, curveWidth, curveAngle, startAngle, centerX, centerY, addition) {
        addition = Math.min(addition, curveRadius);

        const center = new Vector2D(centerX, centerY);

        let innerRadius, innerAngle, innerCenter;

        if (curveAngle === RIGHT_ANGLE) {
            innerRadius = curveRadius;
            innerAngle = curveAngle;
            innerCenter = center.subScalar(addition);
        } else {
            innerRadius = curveRadius - addition;
            innerAngle = enlargeArc(curveAngle, innerRadius, addition * 2);
            innerCenter = center;
        }

        const innerStart = startAngle + (curveAngle - innerAngle) / 2;
        const innerEnd = innerStart + innerAngle;

        const outerRadius = curveRadius + curveWidth + addition;
        const outerAngle = enlargeArc(curveAngle, outerRadius, addition * 2);
        const outerStart = startAngle + (curveAngle - outerAngle) / 2;
        const outerEnd = outerStart + outerAngle;
        const outerCenter = center;

        const p1 = Vector2D.polar(innerRadius, innerStart, innerCenter);
        const p2 = Vector2D.polar(innerRadius, innerEnd, innerCenter);
        const p3 = Vector2D.polar(outerRadius, outerEnd, outerCenter);
        const p4 = Vector2D.polar(outerRadius, outerStart, outerCenter);

        return `${moveTo(p1)} ${arcTo(innerRadius, p2, 0, 1, 0)} ${lineTo(p3)} ${arcTo(outerRadius, p4, 0, 0, 0)} Z`;
    }
</script>

<path
    d={curvedElementPath(radius, width, angle, start, cx, cy, d)}
    {...attributeList(fill, 'fill')}
    {...attributeList(stroke, 'stroke')}
    {transform}
/>
