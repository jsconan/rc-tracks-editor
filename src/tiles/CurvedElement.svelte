<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { getPoint } from '../helpers/point';
    import { polar } from '../helpers/polar';

    let cls = void 0;
    export { cls as class };
    export let style = void 0;
    export let cx = 0;
    export let cy = 0;
    export let width = 1;
    export let radius = 1;
    export let angle = 1;
    export let start = 0;

    const innerRadius = radius;
    const outerRadius = radius + width;
    const end = start + angle;
    const center = getPoint(cx, cy);

    const p1 = polar(innerRadius, start, center);
    const p2 = polar(innerRadius, end, center);
    const p3 = polar(outerRadius, end, center);
    const p4 = polar(outerRadius, start, center);
</script>

<path
    class={cls}
    {style}
    d="M {p1.x} {p1.y}
       A {innerRadius} {innerRadius} 0 0 1 {p2.x} {p2.y}
       L {p3.x} {p3.y}
       A {outerRadius} {outerRadius} 0 0 0 {p4.x} {p4.y}"
/>
