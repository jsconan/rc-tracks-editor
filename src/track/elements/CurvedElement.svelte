<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { Vector2D } from '../../core/models';

    let cls = void 0;
    export { cls as class };
    export let style = void 0;
    export let cx = 0;
    export let cy = 0;
    export let width = 1;
    export let radius = 1;
    export let angle = 1;
    export let start = 0;

    const innerRadius = radius;
    const outerRadius = radius + width;
    const end = start + angle;
    const center = new Vector2D(cx, cy);

    const p1 = Vector2D.polar(innerRadius, start, center);
    const p2 = Vector2D.polar(innerRadius, end, center);
    const p3 = Vector2D.polar(outerRadius, end, center);
    const p4 = Vector2D.polar(outerRadius, start, center);
</script>

<path
    class={cls}
    {style}
    d="M {p1.x} {p1.y}
       A {innerRadius} {innerRadius} 0 0 1 {p2.x} {p2.y}
       L {p3.x} {p3.y}
       A {outerRadius} {outerRadius} 0 0 0 {p4.x} {p4.y}
       Z"
/>
