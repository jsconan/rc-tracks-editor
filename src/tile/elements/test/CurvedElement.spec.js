/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from '@testing-library/svelte';
import { tick } from 'svelte';
import CurvedElement from '../CurvedElement.svelte';

describe('CurvedElement', () => {
    it('renders with default values', () => {
        const { container } = render(CurvedElement);

        expect(container).toMatchSnapshot();
    });

    it('renders with the given parameters', () => {
        const props = {
            cx: 100,
            cy: 150,
            width: 80,
            radius: 50,
            angle: 30,
            start: 60,
            fill: 'red',
            stroke: {
                stroke: 'green',
                'stroke-width': 2
            },
            transform: 'rotate(90)'
        };
        const { container } = render(CurvedElement, { props });

        expect(container).toMatchSnapshot();
    });

    it.each([
        ['cx', { cx: 40 }],
        ['cy', { cy: 40 }],
        ['width', { width: 40 }],
        ['radius', { radius: 40 }],
        ['angle', { angle: 40 }],
        ['start', { start: 40 }],
        ['d', { d: 10 }],
        ['fill', { fill: 'blue' }],
        ['stroke', { stroke: 'blue' }],
        ['stroke-width', { stroke: { 'stroke-width': 3 } }],
        ['transform', { transform: 'rotate(180)' }]
    ])('updates when the parameter %s is modified', async (title, update) => {
        const props = {
            cx: 100,
            cy: 150,
            width: 80,
            radius: 50,
            angle: 30,
            start: 60,
            fill: 'red',
            stroke: {
                stroke: 'green',
                'stroke-width': 2
            },
            transform: 'rotate(90)'
        };
        const rendered = render(CurvedElement, { props });

        await tick();
        rendered.component.$set(update);
        await tick();
        expect(rendered.container).toMatchSnapshot();
    });
});
