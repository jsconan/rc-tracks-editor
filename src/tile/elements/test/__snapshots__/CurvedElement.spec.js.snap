// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurvedElement renders with default values 1`] = `
<body>
  <div>
    <path
      d="M 1 0 A 1 1 0 0 1 0.9998476951563913 0.01745240643728351 L 1.9996953903127825 0.03490481287456702 A 2 2 0 0 0 2 0 Z"
    />
  </div>
</body>
`;

exports[`CurvedElement renders with the given parameters 1`] = `
<body>
  <div>
    <path
      d="M 125 193.30127018922192 A 50 50 0 0 1 100 200 L 100.00000000000001 280 A 130 130 0 0 0 165 262.583302491977 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter angle is modified 1`] = `
<body>
  <div>
    <path
      d="M 125 193.30127018922192 A 50 50 0 0 1 91.31759111665349 199.2403876506104 L 77.42573690329905 278.02500789158705 A 130 130 0 0 0 165 262.583302491977 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter cx is modified 1`] = `
<body>
  <div>
    <path
      d="M 65 193.30127018922192 A 50 50 0 0 1 40 200 L 40.00000000000001 280 A 130 130 0 0 0 105.00000000000001 262.583302491977 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter cy is modified 1`] = `
<body>
  <div>
    <path
      d="M 125 83.30127018922192 A 50 50 0 0 1 100 90 L 100.00000000000001 170 A 130 130 0 0 0 165 152.583302491977 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter d is modified 1`] = `
<body>
  <div>
    <path
      d="M 127.94857298266358 178.61603166465832 A 40 40 0 0 1 90.10384162981907 188.75649686842578 L 90.00850123238898 289.6430089634881 A 140 140 0 0 0 178.4743962363761 265.9386438394733 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter fill is modified 1`] = `
<body>
  <div>
    <path
      d="M 125 193.30127018922192 A 50 50 0 0 1 100 200 L 100.00000000000001 280 A 130 130 0 0 0 165 262.583302491977 Z"
      fill="blue"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter radius is modified 1`] = `
<body>
  <div>
    <path
      d="M 120 184.64101615137753 A 40 40 0 0 1 100 190 L 100.00000000000001 270 A 120 120 0 0 0 160 253.92304845413264 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter start is modified 1`] = `
<body>
  <div>
    <path
      d="M 138.3022221559489 182.13938048432698 A 50 50 0 0 1 117.10100716628344 196.9846310392954 L 144.46261863233696 272.1600407021681 A 130 130 0 0 0 199.58577760546714 233.5623892592501 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter stroke is modified 1`] = `
<body>
  <div>
    <path
      d="M 125 193.30127018922192 A 50 50 0 0 1 100 200 L 100.00000000000001 280 A 130 130 0 0 0 165 262.583302491977 Z"
      fill="red"
      stroke="blue"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter stroke-width is modified 1`] = `
<body>
  <div>
    <path
      d="M 125 193.30127018922192 A 50 50 0 0 1 100 200 L 100.00000000000001 280 A 130 130 0 0 0 165 262.583302491977 Z"
      fill="red"
      stroke-width="3"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter transform is modified 1`] = `
<body>
  <div>
    <path
      d="M 125 193.30127018922192 A 50 50 0 0 1 100 200 L 100.00000000000001 280 A 130 130 0 0 0 165 262.583302491977 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(180)"
    />
  </div>
</body>
`;

exports[`CurvedElement updates when the parameter width is modified 1`] = `
<body>
  <div>
    <path
      d="M 125 193.30127018922192 A 50 50 0 0 1 100 200 L 100 240 A 90 90 0 0 0 145 227.94228634059948 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;
